finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/service/H2Classifiers.scala

```diff
@@ -127,21 +131,29 @@ object H2Classifiers {
    * failures.
    */
   case object NonRetryableServerFailures extends H2Classifier {
-    override val responseClassifier = {
+    override val responseClassifier: PartialFunction[H2ReqRep, ResponseClass] = {
       case H2ReqRep(_, Throw(_) | Return((Responses.Failure()))) => ResponseClass.NonRetryableFailure
     }
-    override val streamClassifier = {
-      case H2ReqRepFrame(_, Throw(_)) => ResponseClass.NonRetryableFailure
-      case H2ReqRepFrame(_, Return((_, Some(Throw(_))))) => ResponseClass.NonRetryableFailure
+    override val streamClassifier: PartialFunction[H2ReqRepFrame, ResponseClass] = {
+      case H2ReqRep(reqRep: H2ReqRep) if responseClassifier.isDefinedAt(reqRep) =>
+        responseClassifier(reqRep)
+      case H2ReqRepFrame(_,
+        Throw(_)
+        | Return((Responses.Failure(), _))
+        | Return((_, Some(Throw(_))))
+        ) => ResponseClass.NonRetryableFailure
       case _ => ResponseClass.Success
     }
   }
 
   case object AllSuccessful extends H2Classifier {
-    override val responseClassifier = {
+    override val responseClassifier: PartialFunction[H2ReqRep, ResponseClass] = {
       case H2ReqRep(_, Throw(_)) => ResponseClass.NonRetryableFailure
+      case H2ReqRep(_, Return(_)) => ResponseClass.Success
```

Comment:
this will match all responses, regardless of if the stream is empty or not.

if the stream is empty, this should be undefined and we should call the stream classifier with `Return(req, Return((rep, None)))` which should fall into the `_ => ResponseClass.Success` case